community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/RecordScanner.java

```diff
@@ -66,15 +98,66 @@ public void run()
             catch ( Exception e )
             {
                 progress.failed( e );
+                throw Exceptions.launderedException( e );
             }
             processor.close();
             progress.done();
         }
     }
 
-    @Override
-    public void stopScanning()
+    public void runParallel()
     {
-        continueScanning = false;
+        Workers<Worker<RECORD>> workers = new Workers<>( getClass().getSimpleName() );
+        ArrayBlockingQueue<RECORD>[] recordQ = new ArrayBlockingQueue[numberOfThreads];
+        for ( int threadId = 0; threadId < numberOfThreads; threadId++ )
+        {
+            recordQ[threadId] = new ArrayBlockingQueue<>( DefaultCacheAccess.DEFAULT_QUEUE_SIZE );
+            workers.start( new Worker<>( recordQ[threadId], processor ) );
+        }
+
+        long recordsPerCPU = (store.maxCount() / numberOfThreads) + 1;
+        cacheAccess.prepareForProcessingOfSingleStore( recordsPerCPU );
+        int[] recsProcessed = new int[numberOfThreads];
+        int qIndex = 0;
+        int entryCount = 0;
+        for ( RECORD record : store )
+        {
+            try
+            {
+                // do a round robin distribution to maintain physical locality
+                recordQ[qIndex++].put( record );
+                qIndex %= numberOfThreads;
+                recsProcessed[qIndex]++;
+            }
+            catch ( InterruptedException e )
+            {
+                Thread.currentThread().interrupt();
+                break;
+            }
+            progress.set( entryCount++ );
+        }
+        progress.done();
+        for ( Worker<RECORD> worker : workers )
+        {
+            worker.done();
+        }
+        workers.awaitAndThrowOnErrorStrict( RuntimeException.class );
+    }
+
+    private static class Worker<RECORD> extends RecordCheckWorker<RECORD>
+    {
+        private final RecordProcessor<RECORD> processor;
+
+        Worker( ArrayBlockingQueue<RECORD> recordsQ, RecordProcessor<RECORD> processor )
+        {
+            super( recordsQ );
+            this.processor = processor;
+        }
+
+        @Override
+        protected void process( RECORD record )
+        {
+            processor.process( record );
+        }
```

Comment:
Can we split this in 3 parts? 1) SequentialRecordScanner, 2) a ParallelRecordScanner and 3) a CompositeRecordScanner that picks the first or the second based on the stage?